subsys/bluetooth/services/bms.c

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/** @file
+ *  @brief LED Button Service (LBS) sample
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/printk.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/bms.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(bt_gatt_bms_srv, CONFIG_BT_GATT_BMS_SRV_LOG_LEVEL);
+
+#define BT_ATT_ERR_OP_CODE_NOT_SUPPORTED	0x80
+#define BT_ATT_ERR_OPERATION_FAILED		0x81
+
+/* Features relevant for LE Transport only */
+enum bt_gatt_bms_srv_le_feature {
+	/* Delete bond of the requesting device */
+	BT_GATT_BMS_SRV_REQ_DEV = BIT(4),
+
+	/* Delete bond of the requesting device with an authorization code */
+	BT_GATT_BMS_SRV_REQ_DEV_AUTH_CODE = BIT(5),
+
+	/* Delete all bonds on the device */
+	BT_GATT_BMS_SRV_ALL_BONDS = BIT(10),
+
+	/* Delete all bonds on the device with an authorization code */
+	BT_GATT_BMS_SRV_ALL_BONDS_AUTH_CODE = BIT(11),
+
+	/* Delete all bonds on the device except for the bond of
+	 * the requesting device
+	 */
+	BT_GATT_BMS_SRV_ALL_EXCEPT_REQ_DEV = BIT(16),
+
+	/* Delete all bonds on the device except for the bond of
+	 * the requesting device with an authorization code
+	 */
+	BT_GATT_BMS_SRV_ALL_EXCEPT_REQ_DEV_AUTH_CODE = BIT(17),
+};
+
+/* Control Point operation codes that are supported on LE transport */
+enum bt_gatt_bms_srv_op {
+	/* Initiates the procedure to delete the bond of
+	 * the requesting device
+	 */
+	BT_GATT_BMS_SRV_OP_DEL_BOND_REQ_DEVICE = 0x03,
+
+	/* Initiates the procedure to delete all bonds on the device */
+	BT_GATT_BMS_SRV_OP_DEL_ALL_BONDS = 0x06,
+
+	/* Initiates the procedure to delete all bonds except for the one of
+	 * the requesting device
+	 */
+	BT_GATT_BMS_SRV_OP_DEL_ALL_BUT_ACTIVE_BOND = 0x09,
+};
+
+/* State of BMS instance */
+enum bt_gatt_bms_srv_state {
+	/* Idle */
+	IDLE,
+
+	/* Control point operation in progress */
+	OPERATION_IN_PROGRESS,
+
+	/* Bond deletion operation in progress */
+	DELETION_IN_PROGRESS,
+};
+
+/* Execution descriptor */
+struct bt_gatt_bms_srv_bond_add_exec {
+	/* Kernel work item */
+	struct k_work work;
+
+	/* Bond operation code */
+	enum bt_gatt_bms_srv_op op_code;
+
+	/* Requestor address */
+	bt_addr_le_t req_addr;
+};
+
+static struct bt_gatt_bms_srv {
+	enum bt_gatt_bms_srv_state state;
+	struct bt_gatt_bms_srv_bond_add_exec bond_add_exec;
+	struct k_work bond_del_exec;
+	struct bt_gatt_bms_srv_cb cbs;
```

Comment:
`struct bt_gatt_bms_srv_cb cbs;` ->`const struct bt_gatt_bms_srv_cb *cbs;`
Maybe it would be better to keeps callback as pointer ?